funflow/src/Control/FunFlow/Exec/Simple.hs

```diff
@@ -94,22 +86,18 @@ runFlowEx _ cfg store runWrapped confIdent flow input = do
       $ AsyncA f
     runFlow' po (External toTask) = AsyncA $ \x -> do
       chash <- contentHash (x, toTask x)
-      submitTask po $ TaskDescription chash (toTask x)
-      KnownTask _ <- awaitTask po chash
-      CS.waitUntilComplete store chash >>= \case
-        Nothing -> fail "Remote process failed to construct item"
-        Just item -> return item
+      CS.constructOrWait store chash >>= \case
+        CS.Pending void -> absurd void
+        CS.Complete item -> return item
+        CS.Missing _ -> do
+          submitTask po $ TaskDescription chash (toTask x)
+          CS.waitUntilComplete store chash >>= \case
+            Nothing -> fail "Remote process failed to construct item"
```

Comment:
Unrelated, but we shouldn't use `fail` here.

Reply:
I've added a corresponding exception type and use `throwM` instead of `fail` now.